src/target/mspm0.c

```diff
@@ -0,0 +1,385 @@
+#define MSPM0_FLASHCTL_CMDTYPE_NOOP      0U
+#define MSPM0_FLASHCTL_CMDTYPE_PROG      1U
+#define MSPM0_FLASHCTL_CMDTYPE_ERASE     2U
+#define MSPM0_FLASHCTL_CMDTYPE_RDVERIFY  3U
+#define MSPM0_FLASHCTL_CMDTYPE_BLVERIFY  6U
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_1WORD  (0U << 4U)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_2WORDS (1U << 4U)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_4WORDS (2U << 4U)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_8WORDS (3U << 4U)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_SECTOR (4U << 4U)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_BANK   (5U << 4U)
+#define MSPM0_FLASHCTL_CMDEXEC_EXEC      1U
+#define MSPM0_FLASHCTL_STAT_DONE         0x01U
+#define MSPM0_FLASHCTL_STAT_CMDPASS      0x02U
+
+typedef struct mspm0_flash {
+	target_flash_s target_flash;
+	uint32_t banks;
+} mspm0_flash_s;
+
+static bool mspm0_flash_erase(target_flash_s *flash, target_addr_t addr, size_t length);
+static bool mspm0_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t length);
+static bool mspm0_mass_erase(target_s *target);
+
+#if MSPM0_CONFIG_FLASH_DUMP_SUPPORT
+static bool mspm0_dump_factory_config(target_s *const target, const int argc, const char **const argv);
+static bool mspm0_dump_bcr_config(target_s *const target, const int argc, const char **const argv);
+
+static command_s mspm0_cmds_list[] = {
+	{"dump_factory", mspm0_dump_factory_config, "Display FACTORY registers"},
+	{"dump_bcr", mspm0_dump_bcr_config, "Display NONMAIN (BCR/BSL) registers"},
+	{NULL, NULL, NULL},
+};
+
+typedef struct conf_register {
+	uint16_t reg_offset;
+	uint16_t size_words;
+	char const *id;
+} conf_register_s;
+
+static conf_register_s mspm0_factory_regs[] = {
+	{0x00U, 1U, "TRACEID"},
+	{0x04U, 1U, "DEVICEID"},
+	{0x08U, 1U, "USERID"},
+	{0x0cU, 1U, "BSLPIN_UART"},
+	{0x10U, 1U, "BSLPIN_I2C"},
+	{0x14U, 1U, "BSLPIN_INVOKE"},
+	{0x18U, 1U, "SRAMFLASH"},
+	{0x3cU, 1U, "TEMP_SENSE0"},
+	{0x7cU, 1U, "BOOTCRC"},
+	{0U, 0U, NULL},
+};
+
+static conf_register_s mspm0_bcr_regs[] = {
+	{0x00U, 1U, "BCRCONFIGID"},
+	{0x04U, 1U, "BOOTCFG0"},
+	{0x08U, 1U, "BOOTCFG1"},
+	{0x0cU, 4U, "PWDDEBUGLOCK"},
+	{0x1cU, 4U, "BOOTCFG2"},
+	{0x20U, 1U, "BOOTCFG3"},
+	{0x24U, 4U, "PWDMASSERASE"},
+	{0x34U, 4U, "PWDFACTORYRESET"},
+	{0x44U, 1U, "FLASHSWP0"},
+	{0x48U, 1U, "FLASHSWP1"},
+	{0x4cU, 1U, "BOOTCFG4"},
+	{0x50U, 1U, "APPCRCSTART"},
+	{0x54U, 1U, "APPCRCLENGTH"},
+	{0x58U, 1U, "APPCRC"},
+	{0x5cU, 1U, "BOOTCRC"},
+	{0x100U, 1U, "BSLCONFIGID"},
+	{0x104U, 1U, "BSLPINCFG0"},
+	{0x108U, 1U, "BSLPINCFG1"},
+	{0x10cU, 1U, "BSLCONFIG0"},
+	{0x110U, 8U, "BSLPW"},
+	{0x130U, 1U, "BSLPLUGINCFG"},
+	{0x134U, 4U, "BSLPLUGINHOOK"},
+	{0x144U, 1U, "PATCHHOOKID"},
+	{0x148U, 1U, "SBLADDRESS"},
+	{0x14cU, 1U, "BSLAPPVER"},
+	{0x150U, 1U, "BSLCONFIG1"},
+	{0x154U, 1U, "BSLCRC"},
+	{0U, 0U, NULL},
+};
+
+static void mspm0_dump_regs(target_s *const target, const conf_register_s *regs, uint32_t base)
+{
+	for (conf_register_s const *r = regs; r->id; ++r) {
```

Comment:
Ditto for this, along with a non-single-letter variable such as `reg` (ie `const conf_register_s *reg = regs;`)